### ISSUE TITLE
Returning focus to a contextual menu is disorienting

### ISSUE THREAD

**Issue by draganescu** [WordPress/gutenberg#54113]
### Description

Recently #53735 was merged providing the feature to rename groups in list view. This was a feature with multiple approaches being tried due to challenging accessibility decisions.

While the renaming currently being done via a modal can be considered a general cognitive common denominator - a modal experience, while interrupting is focused, there is one thing that is not a general cognitive common denominator but very specific to users of assistive technology: focus return to the element that opened the modal dialog.

Because we have the rename action in a contextual menu, once the renaming is confirmed, having the menu open conveys a state of unfinishedness - as if the renaming was not a success. Since the name did change it is very disorienting to see the contextual menu still open - it's a bit like going backwards in time.

In general all contextual menus close when their actions are executed and the focus is moved to the destination affected. In this case we should move the focus after the modal closes to the affected destination which is the list view element whose name has changed.

For users of assistive technology we need to announce that the action was a success and that the focus is now the list item element: the flow of renaming is complete.

### Video of current behavior

https://github.com/WordPress/gutenberg/assets/107534/15fa134b-6286-44d0-b12e-e92b8fcf9b40

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by jeryj** [comment]
From the w3.org dialog pattern explanation, it looks like it's OK to not return focus to the triggering button if it's unlikely that a user would need to retrigger that action:

> When a dialog closes, focus returns to the element that invoked the dialog unless either:
> …
> It is very unlikely users need to immediately re-invoke the dialog.

I think it's very unlikely a user would need to immediately re-invoke a dialog after confirming a rename. But, maybe only after _a successful renaming_. Cancelling seems more likely that it was an accidental invocation, and focus could return to the dropdown menu. Would this start to get even more confusing though? Is returning focus to what triggered the modal so common that a screen reader user would be more likely to feel lost, even if we have a message communicating the successful rename and placing focus on list item?

If the w3.org argument is accurate that it's OK to not send focus to the triggering element in this instance, then in here's a proposed flow:

- Trigger modal via Rename in dropdown 
- **On successful rename:**
  - Send focus to list item option dropdown (three dots) button. This would place it in a similar context while also having the dropdown be closed.
  - Announce a successful rename of the group.
- **On cancel**
  - Send focus back to the Rename dropdown menu item

**Comment by afercia** [comment]
I'd second what @jeryj said above, as in: I'd think the most logial interaction should be different dependin gon success renaming / cancel.

A few thoughts:

> there is one thing that is not a general cognitive common denominator but very specific to users of assistive technology: focus return to the element that opened the modal dialog.

Keyboard, for example, is not an assitive technology. While focus management is necessary for screen reader users, it is necessary for _all_ keyboard usrs. It would be great to switch mindset and start thinking that the we is not a place for only pointing devices or touch. There's a lot of different deivices users can use, and the web has been designed to be device-agnostic.

> Because we have the rename action in a contextual menu, once the renaming is confirmed, having the menu open conveys a state of unfinishedness 

I'd argue the root problem is that a modal dialog should not open from a contextual menu that is expected to close after one of its items is activated. The solution identified above may be good enough but, again, teh root problem is in the design. A separate control to open the modal dialog would have been better.

## More importantly
I see some inconsistency with the focus behavior for other controls in the drop down meny that open modal dialogs as well.
For example
- Create pattern
- Lock

I think the focus management should be the same for all the controls in the contextual drop down menu. RIght now, it is very different. For example:
- Successful Lock: focus is moved back to the menu item, the contextual menu stays open. So this is the same problem as the one in this issue.
- Successful Create pattern: the contextual menu closes and focus is moved to the pattern in the editor content (I think?)

Such interactions should behave in the same way and focus should behave consistently. Seems to me having the focus landing in different places depending on the action taken, with no clear predictability for users, is less than ideal and very confusing.

**Comment by draganescu** [comment]
I agree with @afercia that the whole dot menu should work like a contextual menu cc @jasmussen  - but maybe that should be its own issue. For this issue what Andrea identified is that we have other modals as well whch behave in the disorientating way the OP explains (e.g. lock) - and for them I think the middle ground described by @jeryj is quite good as a solution.

**Comment by jasmussen** [comment]
Not sure what value I can bring to this, I can only agree that it seems disorienting as it is today, and that sending focus back to the ellipsis menu, or even just the list view item itself, on a successful rename seems better.

**Comment by bph** [comment]
This seems to be related to this observation by @andrewserong 

- #54165

**Comment by bph** [comment]
@draganescu do you see this being solved for WP 6.4 before Sept. 25?

**Comment by draganescu** [comment]
Nope, I removed it from the project.

**Comment by andrewserong** [comment]
> This seems to be related to this observation by @andrewserong
> * https://github.com/WordPress/gutenberg/issues/54165

Thanks for the ping @bph! It's similar in nature in the sense that it's about handling focus in the list view in a way that feels intuitive to a user, but I think it's likely a separate issue from the technical side of things.

**Comment by alexstine** [comment]
It's a global problem in Gutenberg. I could understand placing focus on the submenu trigger but not keeping the submenu open. Another example of this is the Options > Preferences modal. When closing the modal, focus should return to Options trigger with menu collapsed.

**Comment by anevins12** [comment]
As desktop keyboard users can dismiss the contextual menu via the ESC key, are we saying that this is not a sufficient method of doing so?

Is the ESC key not a predictable mechanism of dismissing non-modal stuff in WordPress?
